File: src/backend/app/api/v1/endpoints/achievement.py
"""
成就系统 API 端点
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.user import User
from app.services.achievement_service import AchievementService

router = APIRouter()


@router.get("/stats")
async def get_user_stats(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取用户统计数据
    
    返回：
    - total_nodes: 总节点数
    - mastered_nodes: 已掌握节点数
    - total_reviews: 总复习次数
    - total_graphs: 知识图谱数
    - current_streak: 连续学习天数
    - night_reviews: 深夜复习次数
    - morning_reviews: 清晨复习次数
    - perfect_week: 是否完美一周
    """
    service = AchievementService(db)
    stats = await service.get_user_stats(current_user.id)
    
    return {
        "code": 200,
        "message": "获取统计数据成功",
        "data": stats
    }


@router.get("/level")
async def get_user_level(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取用户等级信息
    
    返回：
    - level: 当前等级
    - total_exp: 总经验值
    - current_level_exp: 当前等级起始经验
    - next_level_exp: 下一等级所需经验
    - exp_in_level: 当前等级内的经验
    - exp_to_next: 距离下一等级的经验
    - progress: 当前等级进度（百分比）
    """
    service = AchievementService(db)
    level_info = await service.calculate_level_and_exp(current_user.id)
    
    return {
        "code": 200,
        "message": "获取等级信息成功",
        "data": level_info
    }


@router.get("/achievements")
async def get_user_achievements(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取用户成就列表
    
    返回：
    - unlocked: 已解锁的成就列表
    - locked: 未解锁的成就列表
    - total: 总成就数
    - unlocked_count: 已解锁数量
    - progress: 成就完成度（百分比）
    """
    service = AchievementService(db)
    achievements = await service.get_achievements(current_user.id)
    
    return {
        "code": 200,
        "message": "获取成就列表成功",
        "data": achievements
    }


@router.get("/profile")
async def get_user_profile(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """
    获取用户完整档案
    
    包含：统计数据 + 等级信息 + 成就列表
    """
    service = AchievementService(db)
    profile = await service.get_user_profile(current_user.id)
    
    return {
        "code": 200,
        "message": "获取用户档案成功",
        "data": profile
    }

File: src/frontend/src/services/achievement.ts
/**
 * 成就系统服务
 */

import api from './api';

export interface UserStats {
  total_nodes: number;
  mastered_nodes: number;
  total_reviews: number;
  total_graphs: number;
  current_streak: number;
  night_reviews: number;
  morning_reviews: number;
  perfect_week: boolean;
}

export interface LevelInfo {
  level: number;
  total_exp: number;
  current_level_exp: number;
  next_level_exp: number;
  exp_in_level: number;
  exp_to_next: number;
  progress: number;
}

export interface Achievement {
  id: string;
  name: string;
  description: string;
  icon: string;
  category: string;
  unlocked: boolean;
}

export interface AchievementsData {
  unlocked: Achievement[];
  locked: Achievement[];
  total: number;
  unlocked_count: number;
  progress: number;
}

export interface UserProfile {
  stats: UserStats;
  level: LevelInfo;
  achievements: AchievementsData;
}

/**
 * 获取用户统计数据
 */
export const getUserStats = async (): Promise<UserStats> => {
  const response = await api.get('/api/v1/achievements/stats');
  return response.data.data;
};

/**
 * 获取用户等级信息
 */
export const getUserLevel = async (): Promise<LevelInfo> => {
  const response = await api.get('/api/v1/achievements/level');
  return response.data.data;
};

/**
 * 获取用户成就列表
 */
export const getUserAchievements = async (): Promise<AchievementsData> => {
  const response = await api.get('/api/v1/achievements/achievements');
  return response.data.data;
};

/**
 * 获取用户完整档案
 */
export const getUserProfile = async (): Promise<UserProfile> => {
  const response = await api.get('/api/v1/achievements/profile');
  return response.data.data;
};

/**
 * 获取等级称号
 */
export const getLevelTitle = (level: number): string => {
  if (level >= 20) return '传奇大师';
  if (level >= 18) return '至尊学者';
  if (level >= 16) return '博学鸿儒';
  if (level >= 14) return '知识巨匠';
  if (level >= 12) return '学识渊博';
  if (level >= 10) return '知识大师';
  if (level >= 8) return '学有所成';
  if (level >= 6) return '勤奋学者';
  if (level >= 4) return '知识探索者';
  if (level >= 2) return '初窥门径';
  return '初学者';
};

/**
 * 获取等级颜色
 */
export const getLevelColor = (level: number): string => {
  if (level >= 18) return '#ff4d4f'; // 红色 - 传奇
  if (level >= 15) return '#fa8c16'; // 橙色 - 史诗
  if (level >= 12) return '#a0d911'; // 黄绿 - 稀有
  if (level >= 8) return '#52c41a';  // 绿色 - 优秀
  if (level >= 4) return '#1890ff';  // 蓝色 - 良好
  return '#8c8c8c';                   // 灰色 - 新手
};

/**
 * 获取成就分类名称
 */
export const getAchievementCategoryName = (category: string): string => {
  const categoryMap: Record<string, string> = {
    milestone: '学习里程碑',
    review: '复习成就',
    mastery: '掌握成就',
    streak: '连续学习',
    graph: '知识图谱',
    special: '特殊成就',
  };
  return categoryMap[category] || category;
};

/**
 * 获取成就分类颜色
 */
export const getAchievementCategoryColor = (category: string): string => {
  const colorMap: Record<string, string> = {
    milestone: '#1890ff',
    review: '#52c41a',
    mastery: '#faad14',
    streak: '#ff4d4f',
    graph: '#722ed1',
    special: '#eb2f96',
  };
  return colorMap[category] || '#8c8c8c';
};

